packages/@aws-cdk/aws-rds/lib/instance-engine.ts

```diff
@@ -0,0 +1,269 @@
+   * for example "5.1.42".
+   *
+   * @default - use the default version for this engine type
+   */
+  readonly engineVersion?: string;
+
+  /** The application used by this engine to perform rotation for a single-user scenario. */
+  readonly singleUserRotationApplication: secretsmanager.SecretRotationApplication;
+
+  /** The application used by this engine to perform rotation for a multi-user scenario. */
+  readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
+
+  /**
+   * The default instanceType to use with this engine.
+   */
+  readonly defaultInstanceType: ec2.InstanceType;
+
+  /**
+   * Returns a new instance of this engine with a particular version specified.
+   * Calling {@link engineVersion} on the returned instance will always return the provided value.
+   *
+   * @param engineVersion the exact engine version, for example "5.1.42"
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-rds-database-instance.html#cfn-rds-dbinstance-engineversion
+   */
+  withVersion(engineVersion: string): IInstanceEngine;
+
+  /**
+   * Method called when the engine is used to create a new instance.
+   */
+  bindToInstance(scope: core.Construct, options: InstanceEngineBindOptions): InstanceEngineBindConfig;
+}
+
+abstract class AbstractInstanceEngine implements IInstanceEngine {
+  public readonly engineType: string;
+  public readonly engineVersion?: string;
+  public readonly singleUserRotationApplication: secretsmanager.SecretRotationApplication;
+  public readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
+  public readonly defaultInstanceType: ec2.InstanceType;
+
+  constructor(
+    engineName: string,
+    singleUserRotationApplication: secretsmanager.SecretRotationApplication,
+    multiUserRotationApplication: secretsmanager.SecretRotationApplication,
+    _parameterGroupFamilies: ParameterGroupFamilyMapping[],
+    defaultInstanceType?: ec2.InstanceType,
+    engineVersion?: string) {
+
+    this.engineType = engineName;
+    this.singleUserRotationApplication = singleUserRotationApplication;
+    this.multiUserRotationApplication = multiUserRotationApplication;
+    this.defaultInstanceType = defaultInstanceType ?? ec2.InstanceType.of(ec2.InstanceClass.M5, ec2.InstanceSize.LARGE);
+    this.engineVersion = engineVersion;
+  }
+
+  public withVersion(engineVersion: string): IInstanceEngine {
+    class CopyWithVersion extends AbstractInstanceEngine {}
+
+    return new CopyWithVersion(
+      this.engineType,
+      this.singleUserRotationApplication,
+      this.multiUserRotationApplication,
+      [],
+      this.defaultInstanceType,
+      engineVersion,
+    );
+  }
+
+  public bindToInstance(_scope: core.Construct, _options: InstanceEngineBindOptions): InstanceEngineBindConfig {
+    return {
+      defaultInstanceType: this.defaultInstanceType,
+    };
+  }
+}
+
+class NonSqlServerInstanceEngine extends AbstractInstanceEngine {
```

Comment:
There other properties that are specific to one or more engines. For example `Domain` (not implemented today) is only supported for Oracle and SQL Server. And in the future I can imagine that there will be more properties that are specific to a subset of engines. Is `NonSqlServerInstanceEngine` the right naming/partitioning? How will this evolve?

Reply:
I think this design could handle this quite well. Notice that `NonSqlServerInstanceEngine` in package-private - it's not part of the public API of this module, it's just an implementation detail.

We have other interfaces in this PR, though, that extend `IClusterEngine`, for example `IAuroraClusterEngine`. Right now, it only contains methods like `withVersion_1_17_9()`, but it could be extended with other methods/properties in the future.

Does this make sense?